cloudbuild-k8s.yaml

```diff
@@ -20,35 +25,86 @@ substitutions:
 steps:
 
-- id: Build C2D Docker Image
+- id: Pull Dev Image
   name: gcr.io/cloud-builders/docker
+  waitFor:
+  - '-'
+  dir: k8s
+  entrypoint: bash
   args:
-  - build
-  - --tag
-  - click-to-deploy
-  - --file
-  - k8s.Dockerfile
-  - .
-
-- id: Run Cloud Build K8s Generator
-  name: click-to-deploy
-  entrypoint: python
-  args:
-  - scripts/cloudbuild_k8s_generator.py
-  - --solution
-  - $_SOLUTION_NAME
-  - --output
-  - k8s/$_SOLUTION_NAME/cloudbuild.yaml
+  - -exc
+  - |
+    TAG="$$(cat ./MARKETPLACE_TOOLS_TAG)"
+    docker pull "gcr.io/cloud-marketplace-tools/k8s/dev:$$TAG"
+    docker tag "gcr.io/cloud-marketplace-tools/k8s/dev:$$TAG" "gcr.io/cloud-marketplace-tools/k8s/dev:local"
 
-- id: Test Solution
+- id: Initialize Credentials
   name: gcr.io/cloud-builders/gcloud
+  waitFor:
+  - '-'
+  entrypoint: bash
+  args:
+  - -exc
+  - |
+    gcloud container clusters get-credentials '${_CLUSTER_NAME}' --zone '${_CLUSTER_LOCATION}' --project '$PROJECT_ID'
+    mkdir -p /workspace/.kube/
+    cp -r $$HOME/.kube/ /workspace/
+    mkdir -p /workspace/.config/gcloud/
+    cp -r $$HOME/.config/gcloud/ /workspace/.config/
+
+- id: Install CRDs
+  name: gcr.io/cloud-marketplace-tools/k8s/dev:local
+  env:
+  - 'KUBE_CONFIG=/workspace/.kube'
+  - 'GCLOUD_CONFIG=/workspace/.config/gcloud'
+  - 'EXTRA_DOCKER_PARAMS=--net cloudbuild'
+  dir: k8s/$_SOLUTION_NAME
+  entrypoint: bash
+  args:
+  - -exc
+  - make crd/install
+
+- id: Run diagnostic tool
+  name: gcr.io/cloud-marketplace-tools/k8s/dev:local
+  env:
+  - 'KUBE_CONFIG=/workspace/.kube'
+  - 'GCLOUD_CONFIG=/workspace/.config/gcloud'
+  - 'EXTRA_DOCKER_PARAMS=--net cloudbuild'
+  dir: k8s/$_SOLUTION_NAME
+  entrypoint: bash
+  args:
+  - -exc
+  - make app/doctor
+
+- id: Build Solution
+  name: gcr.io/cloud-marketplace-tools/k8s/dev:local
+  env:
+  - 'KUBE_CONFIG=/workspace/.kube'
+  - 'GCLOUD_CONFIG=/workspace/.config/gcloud'
+  - 'EXTRA_DOCKER_PARAMS=--net cloudbuild'
+  dir: k8s/$_SOLUTION_NAME
+  entrypoint: bash
+  args:
+  - -exc
+  - make -j4 app/build
+
+- id: Verify Solution
+  name: gcr.io/cloud-marketplace-tools/k8s/dev:local
+  env:
+  - 'KUBE_CONFIG=/workspace/.kube'
+  - 'GCLOUD_CONFIG=/workspace/.config/gcloud'
+  - 'EXTRA_DOCKER_PARAMS=--net cloudbuild'
+  dir: k8s/$_SOLUTION_NAME
+  entrypoint: bash
   args:
-  - builds
-  - submit
-  - --timeout
-  - 1800s
-  - --config
-  - k8s/$_SOLUTION_NAME/cloudbuild.yaml
-  - --substitutions
-  - _CLUSTER_NAME=$_CLUSTER_NAME,_CLUSTER_LOCATION=$_CLUSTER_LOCATION
-  - .
+  - -exc
+  - |
+    set -eo pipefail
+    # Export the environment variables passed as json key/value map
+    # Example, --substitutions _EXTRA_VARS='{\"key1\": \"value1\", \"key2\":\"value2\"}' results in:
+    #   export key1=value1
+    #   export key2=value2
+    for extra_var in $(echo $_EXTRA_VARS | jq -r "to_entries|map(\"\(.key)=\(.value|tostring)\")|.[]" ); do
+      export $$extra_var
+    done
+    make -j4 app/verify
```

Comment:
Well, this doesn't do the same things as current configuration, right?

It seems like that `app/verify` is executed only once.

Reply:
Huy and I agreed to merge this and add the extra triggers for wordpress that use the extra parameters

Reply:
It's up to you, but we will stop being able to to test solutions with different configurations.

For example:
1. Default configuration.
2. Enable KALM.
3. Enable Stackdriver Monitoring.
4. Enable Public IP (access).

Reply:
Agreed that it is important to maintain the ability to test multiple configurations for a single application.

The thinking here is to statically declare these configurations directly in the trigger yaml definitions, instead in the code. For example,  the trigger definition for Wordpress will have 3 additional configurations; this will result in a total of 4 triggers being generated for Wordpress.

Reply:
Does cloudbuild-k8s.yaml file includes only one step that execute `make app/verify` only once. How is it going to work with multiple configuration?

Reply:
Do you want to create a trigger for each configuration?

How is this better than the current solution if we start to use multiple configuration with in other solutions, not just WordPress?

1. It will duplicate all pre-steps steps like building image.
2. It will execute more builds than now.
3. We will stop having configuration as a code since Cloud Build cannot read it from the repo yet.

Reply:
@vcanaa Can you summarize what the repo will look like after all the changes are made?

Replies to your specific question:
1. I missed that. It's a good point that the same set of application images (including the deployer) are independently built by each of the build.
2. It's actually less, for the time being, since now each solution requires exactly 2 builds. If each solution has 2+ configurations, then the total number of builds will be indeed more.
3. We will end up having `cloudbuild-*.yaml` files for each configuration materialized in the repo, vs being generated on the fly like currently. Can you clarify what you think the ideal case should be?

Reply:
There will be one cloudbuild-k8s.yaml used by all triggers

Wordpress will have 3 triggers, and it will build the solution 3 times. If we want to avoid, we could still write one cloudbuild for wordpress or other solution that require special config.

IMO, the use case of wordpress does not justify triggering 2 builds for solution, this change is needed before I can proceed.

Reply:
I have two notes here:

1. If you want to have multiple triggers for each solution, e.g. wordpress, then you don't really need `_VERIFY_ENV_VARS` since you can specify the variables into the [trigger config](https://github.com/GoogleCloudPlatform/click-to-deploy/blob/7a75cdc27101918cdd35980d42ede657540e3e60/triggers/k8s/drupal.yaml#L29-L30).
2. If you want to add a new set of configuration, then you will need to do more actions than now. 1) create a config 2) merge it into the repo 3) manually deploy it to cloud build 4) resync if changes are needed.

I agree that triggering multiple builds it's not ideal, and **if and only If** each solution has 2+ configurations, then the total number of builds will be indeed more. But that's what the idea.

Each solution supports prometheus metrics and Stackdriver monitoring. We already have 3 configurations for each solution that should be running, but they aren't.

Ideal solution will be to use local version of cloud build that can execute verification locally, without triggering a new build (but it's not going to work today). 

I am not against this change and I see the benefit of not running two builds each time, but in my opinion, before joining this change, we could contact the Cloud Build team and examine whether local use of cloud builds is possible here.